cli/cmd/proxy/loglevel/command.go

```diff
@@ -0,0 +1,318 @@
+
+// fetchAdminPorts retrieves all admin ports for Envoy Proxies running in a pod given namespace.
+func (l *LogLevelCommand) fetchAdminPorts() (map[string]int, error) {
+	adminPorts := make(map[string]int, 0)
+	pod, err := l.kubernetes.CoreV1().Pods(l.namespace).Get(l.Ctx, l.podName, metav1.GetOptions{})
+	if err != nil {
+		return adminPorts, err
+	}
+
+	connectService, isMultiport := pod.Annotations["consul.hashicorp.com/connect-service"]
+
+	if !isMultiport {
+		// Return the default port configuration.
+		adminPorts[l.podName] = defaultAdminPort
+		return adminPorts, nil
+	}
+
+	for idx, svc := range strings.Split(connectService, ",") {
+		adminPorts[svc] = defaultAdminPort + idx
+	}
+
+	return adminPorts, nil
+}
+
+func (l *LogLevelCommand) fetchLogLevels(adminPorts map[string]int) (map[string]LoggerConfig, error) {
+	loggers := make(map[string]LoggerConfig, 0)
+
+	for name, port := range adminPorts {
+		pf := common.PortForward{
+			Namespace:  l.namespace,
+			PodName:    l.podName,
+			RemotePort: port,
+			KubeClient: l.kubernetes,
+			RestConfig: l.restConfig,
+		}
+
+		logLevels, err := l.logLevelFetcher(l.Ctx, &pf)
+		if err != nil {
+			return loggers, err
+		}
+		loggers[name] = logLevels
+	}
+	return loggers, nil
+}
+
+// FetchLogLevel requests the logging endpoint from Envoy Admin Interface for a given port
+// more can be read about that endpoint https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--logging
+func FetchLogLevel(ctx context.Context, portForward common.PortForwarder) (LoggerConfig, error) {
+	endpoint, err := portForward.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer portForward.Close()
+
+	// this endpoint does not support returning json, so we've gotta parse the plain text
+	response, err := http.Post(fmt.Sprintf("http://%s/logging", endpoint), "text/plain", bytes.NewBuffer([]byte{}))
+	if err != nil {
+		return nil, err
+	}
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to reach envoy: %v", err)
+	}
+	loggers := strings.Split(string(body), "\n")
+	logLevels := make(map[string]string)
+	var name string
+	var level string
+
+	// the first line here is just a header
+	for _, logger := range loggers[1:] {
+		if len(logger) == 0 {
+			continue
+		}
+		fmt.Sscanf(logger, "%s %s", &name, &level)
+		name = strings.TrimRight(name, ":")
+		logLevels[name] = level
+	}
+	return logLevels, nil
+}
+
+func (l *LogLevelCommand) outputLevels(logLevels map[string]LoggerConfig) {
+	l.UI.Output(fmt.Sprintf("Envoy log configuration for %s in namespace default:", l.podName))
+	for n, levels := range logLevels {
+		l.UI.Output(fmt.Sprintf("Log Levels for %s", n), terminal.WithHeaderStyle())
+		table := terminal.NewTable("Name", "Level")
+		for name, level := range levels {
+			table.AddRow([]string{name, level}, []string{"", levelToColor[level]})
+		}
+		l.UI.Table(table)
+		l.UI.Output("")
+	}
+}
+
+func (l *LogLevelCommand) Help() string {
+	l.once.Do(l.init)
```

Comment:
Wondering what the call to `init` here is for? Is `Help` ever called without first calling `Run`? If so, maybe consider adding a `New` method that initializes the struct and calls `init` rather than the bare initialization in `commands.go`

Reply:
AFAIK it can be called before `Run`, I kept this pattern with the `init` to match the rest of the codebase tbh, I'd typically prefer to use a `New` command to ensure everything is initialized first prior to running but went this route to keep it in line with the rest of the repo

Reply:
Yeah, it's a quirk of this CLI library. Help can be called without first calling Run.

Reply:
so if we moved to a constructor and called that in the `commands.go` rather than just initializing the struct we _could_ handle that, the issue would be that we'd be instantiating all of our deps/connections even if this command wasn't called which could slow down usage for other commands (unless the CLI lazily inits the commands, which it kinda looks like it does)

Reply:
nevermind what I said about slowing down initialization with deps, I was mistaken in thinking we could have access to the args and pass them to the constructor but it doesn't look like that's the case so we still init the deps in `Run` but are able to push creating the flags into a constructor, I put together what it would look like [here](https://github.com/hashicorp/consul-k8s/commit/dd0c3618690b3035b94cb3009a1da62897026816), the only drawback is that we'd be deviating from the pattern in the rest of the cli